# Cold Start Neural Architecture Search Example: EfficientNetB7-ResidualMLP using hyperband (on CIFAR10 dataset): Best Model Evaluation:

## By cold start I mean the search does not rely on a memory of or aggregate data from a history of past NAS projects.

1. Ran: task_trigger.py (seen in next cell)
2. The process task_trigger.py parsed and executed the shell script: 2022-02-09_04-29_CIFAR10_EfficientNetB7-ResidualMLP_NAS_SECOND_PASS_SEARCH_task_trigger.sh (The next cell)
3. The shell script passed the configured parameters to and executed task.py, our Neural Architecture Search and training task.
4. This used my python package > module > Class > method: residualmlp.residual_mlp.ResidualMLP.build_auto_residual_mlp. This module is seen in the next cell. The method .build_auto_residual_mlp abstracts away and parameterizes the ResidualMLP neural archtiecture into a selection from a single ordered list of positive integers, in ascending order by model complexity. The ResidualMLP architecture is MLP architecture with strategically arranged skip connections and other nonlinear paths through a series Dense layers. This method described is used to to recursively create tandem EfficientNetB7-ResidualMLP via an integration with the hyperparameter optimization engine keras_tuner (this is done in in task.py) which runs a Hyperband neural architecture search and hyperparameter optimization.

## Here, we evaluate the best model returned from the successful trials of training / search


## task_trigger.py (First step in the cascade)

In [1]:
!cat task_trigger.py

"""Runs task.py, is a more convenient way to set up the parameters than
manually writing a shell script."""

import subprocess
preliminary_install_commands = ["pip3 install --upgrade pip",
                                "pip3 install pendulum"]
for cmd in preliminary_install_commands:
    subprocess.run(cmd,
                   shell=True,
                   check=True)
import pendulum

# Configure the run with this dict.
# Always enter boolean and floats as strings.

TIME_STAMP = pendulum.now().\
                    	__str__().\
                    	replace('T','_').\
                    	replace(':','-')[:16]

HPARAMS = {
    'project_name':"CIFAR10_EfficientNetB7-ResidualMLP_NAS_SECOND_PASS_SEARCH",
    'training_set_size':50000,
    'patience':25,
    'patience_min_delta':0.00001,
    'batch_size':50,
    'max_epochs':150,
    'minimum_learning_rate':0.00007,
    'maximum_learning_rate':0.7,
    'number_of_learning_rates_to_try':7,
    'minimum_number_of_blocks':1,
    'maximum_num

In [2]:
!cat 2022-02-09_04-29_CIFAR10_EfficientNetB7-ResidualMLP_NAS_SECOND_PASS_SEARCH_task_trigger.sh

python3 task.py \
    --project_name 'CIFAR10_EfficientNetB7-ResidualMLP_NAS_SECOND_PASS_SEARCH'\
    --training_set_size '50000'\
    --patience '25'\
    --patience_min_delta '1e-05'\
    --batch_size '50'\
    --max_epochs '150'\
    --minimum_learning_rate '7e-05'\
    --maximum_learning_rate '0.7'\
    --number_of_learning_rates_to_try '7'\
    --minimum_number_of_blocks '1'\
    --maximum_number_of_blocks '5'\
    --minimum_number_of_layers_per_block '2'\
    --maximum_number_of_layers_per_block '4'\
    --minimum_neurons_per_block_layer '60'\
    --maximum_neurons_per_block_layer '110'\
    --n_options_of_neurons_per_layer_to_try '7'\
    --minimum_neurons_per_block_layer_decay '10'\
    --maximum_neurons_per_block_layer_decay '40'\
    --minimum_dropout_rate_for_bypass_layers '0.2'\
    --maximim_dropout_rate_for_bypass_layers '0.6'\
    --n_options_dropout_rate_for_bypass_layers '3'\
    --minimum_inter_block_layers_per_block '0'\
    --maximum_inter_block_layers_per_block '85

In [3]:
!cat task.py

import os
import subprocess
import argparse
# Add parser for common params

try:
    import pendulum
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pendulum", 
    	           shell=True, 
    	           check=True)
    import pendulum

try:
    import pandas as pd
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pandas", 
    	           shell=True, 
    	           check=True)
    import pandas as pd

try:
    import keras_tuner as kt
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install keras_tuner", 
    	           shell=True, 
    	           check=True)
    import keras_tuner as kt


try:
    import tensorflow as

In [4]:
!cat residualmlp/residual_mlp.py

try:
    import keras_tuner as kt
except Exception as exc:
    print("Importing Keras tuner appears to be unsuccesful. "
          "keras tuner may need to be installed $ pip install -q -U "
          "keras-tuner. The auto-ml features are disabled until this is "
           "fixed, but ResidualMLP will work. A more detailed error is: "
           f"{exc}")
import tensorflow as tf
import numpy as np
import pandas as pd
import pendulum

# Becomes a layer to convert BW images to RGB

class ResidualMLP:
    
    def __init__(self, problem_type = 'classification',
                      learning_rate = .0007,
                      minimum_learning_rate = 0.00007,
                      maximum_learning_rate = 0.7,
                      number_of_learning_rates_to_try = 5,
                      input_shape = (32,32,3),
                      bw_images = False,
                      base_model = '',
                      base_model_input_shape = (600,600,3),
                      base_model_hyp

In [5]:
## Head of the console log from the NAS run

In [6]:
!cat 2022-02-09_04-29_CIFAR10_EfficientNetB7-ResidualMLP_NAS_SECOND_PASS_SEARCH_python3_shell_log.txt | head -n 100

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.7/11.7 MB 71.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.0/98.0 KB 16.9 MB/s eta 0:00:00
We're missing a module:
No module named 'pandas'
No problem, we install it...
We're missing a module:
No module named 'keras_tuner'
No problem, we install it...
Date: 2022-02-09_04_29
project_name: CIFAR10_EfficientNetB7-ResidualMLP_NAS_SECOND_PASS_SEARCH
training_set_size: 50000
patience: 25
PATIENCE_MIN_DELTA: 1e-05
BATCH_SIZE: 50
MAX_EPOCHS: 150
RESULTS_DIR_FOR_SEARCH: 2022-02-09_04_29_CIFAR10_EfficientNetB7-ResidualMLP_NAS_SECOND_PASS_SEARCH_SEARCH_RUN
MINIMUM_LEARNING_RATE 7e-05
MAXIMUM_LEARNING_RATE: 0.7
NUMBER_OF_LEARNING_RATES_TO_TRY: 7
MINIMUM_NUMBER_OF_BLOCKS: 1
MAXIMUM_NUMBER_OF_BLOCKS: 5
MINIMUM_NUMBER_OF_LAYERS_PER_BLOCK: 2
MAXIMUM_NUMBER_OF_LAYERS_PER_BLOCK: 4
MINIMUM_NEURONS_PER_BLOCK_LAYER: 60
MAXIMUM_NEURONS_PER_BLOCK_LAYER: 110
N_OPTIONS_OF_NEURONS_PER_LAYER_TO_TRY: 7
MINIMUM_NEURONS_PER_BLOCK_LAYER_DECAY: 10

## Recreate the environment in the task needed to run the evaluation:

In [7]:
! pip3 install matplotlib
! pip3 install numpy
import os
import subprocess
import argparse
# Add parser for common params

try:
    import pendulum
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pendulum", 
    	           shell=True, 
    	           check=True)
    import pendulum

try:
    import pandas as pd
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pandas", 
    	           shell=True, 
    	           check=True)
    import pandas as pd

try:
    import keras_tuner as kt
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install keras_tuner", 
    	           shell=True, 
    	           check=True)
    import keras_tuner as kt


try:
    import tensorflow as tf
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install tensorflow", 
    	           shell=True, 
    	           check=True)
    import tensorflow as tf

!pip3 install matplotlib
import matplotlib.pyplot as plt
    
from residualmlp.residual_mlp import ResidualMLP

     |████████████████████████████████| 11.3 MB 12.4 MB/s eta 0:00:01
     |████████████████████████████████| 1.2 MB 29.7 MB/s eta 0:00:01
     |████████████████████████████████| 4.3 MB 38.0 MB/s eta 0:00:01
     |████████████████████████████████| 895 kB 34.2 MB/s eta 0:00:01
You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.
You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.
We're missing a module:
No module named 'pendulum'
No problem, we install it...


You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.


We're missing a module:
No module named 'pandas'
No problem, we install it...


You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.


We're missing a module:
No module named 'keras_tuner'
No problem, we install it...


You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.


You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.


## Load the training and test set.

In [8]:

cifar = tf.keras.datasets.cifar10.load_data()
(x_train, y_train), (x_test, y_test) = cifar

y_train_ohe = tf.one_hot([i[0] for i in  y_train],10)
y_test_ohe = tf.one_hot([i[0] for i in y_test],10)


170508288/170498071 [==============================] - 8s 0us/step


2022-02-10 05:40:09.507196: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-10 05:40:09.684874: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-10 05:40:09.685948: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-10 05:40:09.692044: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-10 05:40:09.693161: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read f

## Set up a few global parameters for our model

In [9]:
DATE = pendulum.now().__str__()[:16].replace("T","_").replace(":","_")

# Keras tuner search & fit args

PATIENCE = 25
PATIENCE_MIN_DELTA = .00001
BATCH_SIZE = 50
MAX_EPOCHS = 150



# Base model args
INPUT_SHAPE = (32,32,3)
BASE_MODEL_INPUT_SHAPE = (600,600,3)

PROJECT_NAME = "CIFAR10_EfficientNetB7-ResidualMLP_NAS_SECOND_PASS_SEARCH"
TRAINING_SET_SIZE = 50000
PATIENCE = 25
PATIENCE_MIN_DELTA: 1e-05
BATCH_SIZE = 50
MAX_EPOCHS = 150
RESULTS_DIR_FOR_SEARCH =\
    "2022-02-09_04_29_CIFAR10_EfficientNetB7-ResidualMLP_NAS_SECOND_PASS_SEARCH_SEARCH_RUN"
MINIMUM_LEARNING_RATE = 7e-05
MAXIMUM_LEARNING_RATE = 0.7
NUMBER_OF_LEARNING_RATES_TO_TRY = 7
MINIMUM_NUMBER_OF_BLOCKS = 1
MAXIMUM_NUMBER_OF_BLOCKS = 5
MINIMUM_NUMBER_OF_LAYERS_PER_BLOCK = 2
MAXIMUM_NUMBER_OF_LAYERS_PER_BLOCK = 4
MINIMUM_NEURONS_PER_BLOCK_LAYER = 60
MAXIMUM_NEURONS_PER_BLOCK_LAYER = 110
N_OPTIONS_OF_NEURONS_PER_LAYER_TO_TRY = 7
MINIMUM_NEURONS_PER_BLOCK_LAYER_DECAY = 10
MAXIMUM_NEURONS_PER_BLOCK_LAYER_DECAY = 40
MINIMUM_DROPOUT_RATE_FOR_BYPASS_LAYERS = 0.2
MAXIMIM_DROPOUT_RATE_FOR_BYPASS_LAYERS = 0.6
N_OPTIONS_DROPOUT_RATE_FOR_BYPASS_LAYERS = 3
MINIMUM_INTER_BLOCK_LAYERS_PER_BLOCK = 0
MAXIMUM_INTER_BLOCK_LAYERS_PER_BLOCK = 85
N_OPTIONS_INTER_BLOCK_LAYERS_PER_BLOCK = 7
MINIMUM_DROPOUT_RATE = 0.2
MAXIMUM_DROPOUT_RATE = 0.4
N_OPTIONS_DROPOUT_RATE = 5
MINIMUM_FINAL_DENSE_LAYERS = 0
MAXIMUM_FINAL_DENSE_LAYERS = 200
N_OPTIONS_FINAL_DENSE_LAYERS = 7

## Set up our base model: 
1. EfficientNetB7
2. Pre-trained on ImageNet.
3. Set only the last Conv2D layer to trainable.
4. Drop the trailing Dense layer and BatchNormalization layers.

In [10]:
mod_with_fc_raw = tf.keras.applications.efficientnet.EfficientNetB7(
    include_top=True, weights='imagenet', input_tensor=None,
    input_shape = BASE_MODEL_INPUT_SHAPE, pooling='max', classes=1000
)

# Make the deepest conv2d layer trainable, leave everything else
# as not trainable
for layer in mod_with_fc_raw.layers:
    layer.trainable = False
# Last conv2d layer. This we want to train .
mod_with_fc_raw.layers[-6].trainable = True

# Create the final base model
# (remove the final Dense and BatchNormalization layers ...) 
efficient_net_b_7_transferable_base_model =\
    tf.keras.Model(inputs=mod_with_fc_raw.layers[0].input, 
                    outputs=mod_with_fc_raw.layers[-3].output)

268337152/268326632 [==============================] - 12s 0us/step


## Instantiate  the ResidualMLP class object:

In [11]:

model_builder = ResidualMLP(
                    problem_type= "classification",  
                    minimum_learning_rate = MINIMUM_LEARNING_RATE, 
                    maximum_learning_rate = MAXIMUM_LEARNING_RATE, 
                    number_of_learning_rates_to_try =
                        NUMBER_OF_LEARNING_RATES_TO_TRY, 
                    input_shape = (32,32,3), 
                    bw_images = False, 
                    base_model = 
                        efficient_net_b_7_transferable_base_model, 
                    base_model_input_shape = BASE_MODEL_INPUT_SHAPE, 
                    flatten_after_base_model = False, 
                    minimum_number_of_blocks = MINIMUM_NUMBER_OF_BLOCKS, 
                    maximum_number_of_blocks = MAXIMUM_NUMBER_OF_BLOCKS, 
                    minimum_number_of_layers_per_block =
                        MINIMUM_NUMBER_OF_LAYERS_PER_BLOCK, 
                    maximum_number_of_layers_per_block =
                        MAXIMUM_NUMBER_OF_LAYERS_PER_BLOCK,
                    minimum_neurons_per_block_layer =
                        MINIMUM_NEURONS_PER_BLOCK_LAYER, 
                    maximum_neurons_per_block_layer =
                        MAXIMUM_NEURONS_PER_BLOCK_LAYER, 
                    n_options_of_neurons_per_layer_to_try =
                        N_OPTIONS_OF_NEURONS_PER_LAYER_TO_TRY, 
                    minimum_neurons_per_block_layer_decay =
                        MINIMUM_NEURONS_PER_BLOCK_LAYER_DECAY, 
                    maximum_neurons_per_block_layer_decay = 
                        MAXIMUM_NEURONS_PER_BLOCK_LAYER_DECAY, 
                    minimum_dropout_rate_for_bypass_layers =
                        MINIMUM_DROPOUT_RATE_FOR_BYPASS_LAYERS, 
                    maximim_dropout_rate_for_bypass_layers =
                        MAXIMIM_DROPOUT_RATE_FOR_BYPASS_LAYERS, 
                    n_options_dropout_rate_for_bypass_layers =
                        N_OPTIONS_DROPOUT_RATE_FOR_BYPASS_LAYERS,
                    minimum_inter_block_layers_per_block =
                        MINIMUM_INTER_BLOCK_LAYERS_PER_BLOCK, 
                    maximum_inter_block_layers_per_block =
                        MAXIMUM_INTER_BLOCK_LAYERS_PER_BLOCK,
                    n_options_inter_block_layers_per_block =\
                        N_OPTIONS_INTER_BLOCK_LAYERS_PER_BLOCK,
                    minimum_dropout_rate = MINIMUM_DROPOUT_RATE, 
                    maximum_dropout_rate = MAXIMUM_DROPOUT_RATE,
                    n_options_dropout_rate = N_OPTIONS_DROPOUT_RATE, 
                    minimum_final_dense_layers =
                        MINIMUM_FINAL_DENSE_LAYERS,
                    maximum_final_dense_layers =
                        MAXIMUM_FINAL_DENSE_LAYERS, 
                    n_options_final_dense_layers =
                        N_OPTIONS_FINAL_DENSE_LAYERS, 
                    number_of_classes = 10,
                    final_activation = tf.keras.activations.softmax)


## Instantiate our Hyperband object. This will pull the "Oracles" from the default location and restore the object to its state at the end of the training run. 

In [12]:

RESULTS_DIR_FOR_SEARCH =\
    f'{DATE}_{PROJECT_NAME}_SEARCH_RUN'

logdir_for_search = os.path.join("logs", RESULTS_DIR_FOR_SEARCH + "_TB")
tensorboard_callback_search =\
    tf.keras.callbacks.TensorBoard(logdir_for_search, histogram_freq=1)

tuner = kt.Hyperband(
    model_builder.build_auto_residual_mlp,
    objective='val_loss',
    max_epochs = MAX_EPOCHS,
    hyperband_iterations = 2)


2022-02-10 05:40:30.408545: I tensorflow/core/profiler/lib/profiler_session.cc:131] Profiler session initializing.
2022-02-10 05:40:30.408625: I tensorflow/core/profiler/lib/profiler_session.cc:146] Profiler session started.
2022-02-10 05:40:30.410607: I tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1614] Profiler found 1 GPUs


INFO:tensorflow:Reloading Oracle from existing project ./untitled_project/oracle.json


2022-02-10 05:40:30.600298: I tensorflow/core/profiler/lib/profiler_session.cc:164] Profiler session tear down.
2022-02-10 05:40:30.600545: I tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1749] CUPTI activity buffer flushed


All permutations:
      number_of_blocks  layers_per_block  neurons_per_block_layer  \
0                    1                 2                       60   
1                    1                 2                       60   
2                    1                 2                       60   
3                    1                 2                       60   
4                    1                 2                       60   
...                ...               ...                      ...   
3250                 5                 4                      110   
3251                 5                 4                      110   
3252                 5                 4                      110   
3253                 5                 4                      110   
3254                 5                 4                      110   

      neurons_per_block_layer_decay  
0                                10  
1                                11  
2                                12  
3


## Let's look at a summary of the training run's 10 best trials:


In [13]:
tuner.results_summary(10)

Results summary
Results in ./untitled_project
Showing 10 best trials
Objective(name='val_loss', direction='min')
Trial summary
Hyperparameters:
learning_rate: 0.0037513129240040457
blocks: 493
bypass_layers_units: 110
inter_block_layers: 0
final_dense_layers: 133
b_norm_or_dropout_residual_bypass_layers: bnorm
dropout_rate_for_bypass_layers: 0.4
b_norm_or_dropout_last_layers: bnorm
dropout_rate: 0.30000000000000004
tuner/epochs: 2
tuner/initial_epoch: 0
tuner/bracket: 4
tuner/round: 0
Score: 0.2851129472255707
Trial summary
Hyperparameters:
learning_rate: 0.0006246016228043756
blocks: 1498
bypass_layers_units: 85
inter_block_layers: 0
final_dense_layers: 200
b_norm_or_dropout_residual_bypass_layers: dropout
dropout_rate_for_bypass_layers: 0.4
b_norm_or_dropout_last_layers: dropout
dropout_rate: 0.35000000000000003
tuner/epochs: 2
tuner/initial_epoch: 0
tuner/bracket: 4
tuner/round: 0
Score: 0.3170013427734375
Trial summary
Hyperparameters:
learning_rate: 0.0008470649536294128
blocks: 1

## Let's restore the best model: (This has only been trained for 2 epochs)

In [14]:
best_model_2022_02_09 = tuner.get_best_models(num_models=1)[0]

All permutations:
      number_of_blocks  layers_per_block  neurons_per_block_layer  \
0                    1                 2                       60   
1                    1                 2                       60   
2                    1                 2                       60   
3                    1                 2                       60   
4                    1                 2                       60   
...                ...               ...                      ...   
3250                 5                 4                      110   
3251                 5                 4                      110   
3252                 5                 4                      110   
3253                 5                 4                      110   
3254                 5                 4                      110   

      neurons_per_block_layer_decay  
0                                10  
1                                11  
2                                12  
3

In [15]:
best_model_2022_02_09.summary()

Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLayer)            [(None, 32, 32, 3)]  0                                            
__________________________________________________________________________________________________
resizing (Resizing)             (None, 600, 600, 3)  0           input_1[0][0]                    
__________________________________________________________________________________________________
model (Functional)              (None, 2560)         64097687    resizing[0][0]                   
__________________________________________________________________________________________________
dense (Dense)                   (None, 85)           217685      model[1][0]                      
______________________________________________________________________________________________

## Out of the box model from the NAS tested:

In [16]:

best_model_2022_02_09.evaluate(x_test,y_test_ohe)


2022-02-10 05:40:54.088724: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 30720000 exceeds 10% of free system memory.
2022-02-10 05:40:54.129476: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)
2022-02-10 05:41:02.748672: I tensorflow/stream_executor/cuda/cuda_dnn.cc:381] Loaded cuDNN version 8204


313/313 [==============================] - 450s 1s/step - loss: 0.2970 - top_1_categorical_accuracy: 0.9051 - top_2_categorical_accuracy: 0.9680 - top_3_categorical_accuracy: 0.9838 - top_4_categorical_accuracy: 0.9915 - top_5_categorical_accuracy: 0.9959 - top_6_categorical_accuracy: 0.9980 - top_7_categorical_accuracy: 0.9989 - top_8_categorical_accuracy: 0.9994 - top_9_categorical_accuracy: 0.9998 - precision: 0.9228 - recall: 0.8930 - accuracy: 0.0000e+00


[0.29702749848365784,
 0.9050999879837036,
 0.9679999947547913,
 0.9837999939918518,
 0.9915000200271606,
 0.9958999752998352,
 0.9980000257492065,
 0.9988999962806702,
 0.9994000196456909,
 0.9998000264167786,
 0.9228066802024841,
 0.8930000066757202,
 0.0]

## The NAS generated and trained model's performance: top_1_categorical_accuracy: 0.9051

In [17]:
best_hp = tuner.get_best_hyperparameters(1)[0]
final_model = model_builder.build_auto_residual_mlp(best_hp)

All permutations:
      number_of_blocks  layers_per_block  neurons_per_block_layer  \
0                    1                 2                       60   
1                    1                 2                       60   
2                    1                 2                       60   
3                    1                 2                       60   
4                    1                 2                       60   
...                ...               ...                      ...   
3250                 5                 4                      110   
3251                 5                 4                      110   
3252                 5                 4                      110   
3253                 5                 4                      110   
3254                 5                 4                      110   

      neurons_per_block_layer_decay  
0                                10  
1                                11  
2                                12  
3

In [18]:
final_model.summary()

Model: "model_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_2 (InputLayer)            [(None, 32, 32, 3)]  0                                            
__________________________________________________________________________________________________
resizing_1 (Resizing)           (None, 600, 600, 3)  0           input_2[0][0]                    
__________________________________________________________________________________________________
model (Functional)              (None, 2560)         64097687    resizing_1[0][0]                 
__________________________________________________________________________________________________
dense_10 (Dense)                (None, 85)           217685      model[2][0]                      
____________________________________________________________________________________________

## Train the model manually:

In [20]:

RESULTS_DIR_FOR_FINAL_MODEL = "2022-02-09_manually_trained_fianl_model"

logdir_final_model = os.path.join("logs",
                                  RESULTS_DIR_FOR_FINAL_MODEL + "_TB")
tensorboard_callback_final =\
    tf.keras.callbacks.TensorBoard(logdir_final_model, histogram_freq=1)


2022-02-10 05:53:07.141442: I tensorflow/core/profiler/lib/profiler_session.cc:131] Profiler session initializing.
2022-02-10 05:53:07.141516: I tensorflow/core/profiler/lib/profiler_session.cc:146] Profiler session started.
2022-02-10 05:53:08.072859: I tensorflow/core/profiler/lib/profiler_session.cc:164] Profiler session tear down.
2022-02-10 05:53:08.073124: I tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1749] CUPTI activity buffer flushed


In [21]:

history = final_model.fit(x=x_train,  
                    y=y_train_ohe, 
                    batch_size=BATCH_SIZE, 
                    epochs=150,      
                    verbose='auto', 
                    callbacks=[tf.keras.callbacks.\
                               EarlyStopping(monitor='val_loss',
                                             patience=PATIENCE,
                                             min_delta=PATIENCE_MIN_DELTA,
                                             restore_best_weights=True),
                            tensorboard_callback_final], 
                    validation_split=0.1,
                    validation_data=None,
                    shuffle=True,
                    class_weight=None, 
                    sample_weight=None, 
                    initial_epoch=0, 
                    steps_per_epoch=None, 
                    validation_steps=None, 
                    validation_batch_size=50, 
                    validation_freq=1, 
                    max_queue_size=10, 
                    workers=5, 
                    use_multiprocessing=True)


2022-02-10 05:54:13.776717: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 138240000 exceeds 10% of free system memory.
2022-02-10 05:54:13.856200: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 138240000 exceeds 10% of free system memory.
/usr/local/lib/python3.8/dist-packages/keras/utils/generic_utils.py:494: CustomMaskWarning: Custom mask layers require a config and must override get_config. When loading, the custom mask layer must be passed to the custom_objects argument.
  warnings.warn('Custom mask layers require a config and must override '


Epoch 1/150
  1/900 [..............................] - ETA: 6:01:48 - loss: 3.2705 - top_1_categorical_accuracy: 0.1200 - top_2_categorical_accuracy: 0.2000 - top_3_categorical_accuracy: 0.3200 - top_4_categorical_accuracy: 0.3800 - top_5_categorical_accuracy: 0.4400 - top_6_categorical_accuracy: 0.4800 - top_7_categorical_accuracy: 0.5800 - top_8_categorical_accuracy: 0.6800 - top_9_categorical_accuracy: 0.8600 - precision_1: 0.1000 - recall_1: 0.0200 - accuracy: 0.0000e+00

2022-02-10 05:54:39.188578: I tensorflow/core/profiler/lib/profiler_session.cc:131] Profiler session initializing.
2022-02-10 05:54:39.188675: I tensorflow/core/profiler/lib/profiler_session.cc:146] Profiler session started.


  2/900 [..............................] - ETA: 1:02:40 - loss: 2.7756 - top_1_categorical_accuracy: 0.1700 - top_2_categorical_accuracy: 0.3300 - top_3_categorical_accuracy: 0.4200 - top_4_categorical_accuracy: 0.5300 - top_5_categorical_accuracy: 0.6000 - top_6_categorical_accuracy: 0.6800 - top_7_categorical_accuracy: 0.7600 - top_8_categorical_accuracy: 0.8100 - top_9_categorical_accuracy: 0.9100 - precision_1: 0.2083 - recall_1: 0.0500 - accuracy: 0.0000e+00

2022-02-10 05:54:42.550361: I tensorflow/core/profiler/lib/profiler_session.cc:66] Profiler session collecting data.
2022-02-10 05:54:42.551014: I tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1749] CUPTI activity buffer flushed
2022-02-10 05:54:42.655734: I tensorflow/core/profiler/internal/gpu/cupti_collector.cc:673]  GpuTracer has collected 2871 callback api events and 2868 activity events. 
2022-02-10 05:54:42.777403: I tensorflow/core/profiler/lib/profiler_session.cc:164] Profiler session tear down.
2022-02-10 05:54:42.932690: I tensorflow/core/profiler/rpc/client/save_profile.cc:136] Creating directory: logs/2022-02-09_manually_trained_fianl_model_TB/train/plugins/profile/2022_02_10_05_54_42

2022-02-10 05:54:42.981485: I tensorflow/core/profiler/rpc/client/save_profile.cc:142] Dumped gzipped tool data for trace.json.gz to logs/2022-02-09_manually_trained_fianl_model_TB/train/plugins/profile/2022_02_10_05_54_42/nqnzthuxgu.trace.json.gz
2022-02-10 05:54:43.170175: I tensor

900/900 [==============================] - 2293s 3s/step - loss: 0.3827 - top_1_categorical_accuracy: 0.8817 - top_2_categorical_accuracy: 0.9554 - top_3_categorical_accuracy: 0.9763 - top_4_categorical_accuracy: 0.9865 - top_5_categorical_accuracy: 0.9921 - top_6_categorical_accuracy: 0.9954 - top_7_categorical_accuracy: 0.9969 - top_8_categorical_accuracy: 0.9981 - top_9_categorical_accuracy: 0.9992 - precision_1: 0.9062 - recall_1: 0.8606 - accuracy: 0.0000e+00 - val_loss: 0.2796 - val_top_1_categorical_accuracy: 0.9108 - val_top_2_categorical_accuracy: 0.9702 - val_top_3_categorical_accuracy: 0.9858 - val_top_4_categorical_accuracy: 0.9936 - val_top_5_categorical_accuracy: 0.9972 - val_top_6_categorical_accuracy: 0.9990 - val_top_7_categorical_accuracy: 0.9998 - val_top_8_categorical_accuracy: 1.0000 - val_top_9_categorical_accuracy: 1.0000 - val_precision_1: 0.9245 - val_recall_1: 0.8990 - val_accuracy: 0.0000e+00


2022-02-10 06:32:30.167044: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 36126720 exceeds 10% of free system memory.
2022-02-10 06:32:30.250268: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 36126720 exceeds 10% of free system memory.


Epoch 2/150
900/900 [==============================] - 2216s 2s/step - loss: 0.2881 - top_1_categorical_accuracy: 0.9079 - top_2_categorical_accuracy: 0.9698 - top_3_categorical_accuracy: 0.9855 - top_4_categorical_accuracy: 0.9926 - top_5_categorical_accuracy: 0.9963 - top_6_categorical_accuracy: 0.9982 - top_7_categorical_accuracy: 0.9990 - top_8_categorical_accuracy: 0.9996 - top_9_categorical_accuracy: 0.9998 - precision_1: 0.9243 - recall_1: 0.8935 - accuracy: 0.0000e+00 - val_loss: 0.2676 - val_top_1_categorical_accuracy: 0.9178 - val_top_2_categorical_accuracy: 0.9728 - val_top_3_categorical_accuracy: 0.9874 - val_top_4_categorical_accuracy: 0.9952 - val_top_5_categorical_accuracy: 0.9980 - val_top_6_categorical_accuracy: 0.9990 - val_top_7_categorical_accuracy: 0.9994 - val_top_8_categorical_accuracy: 0.9996 - val_top_9_categorical_accuracy: 1.0000 - val_precision_1: 0.9286 - val_recall_1: 0.9084 - val_accuracy: 0.0000e+00
Epoch 3/150
 26/900 [..............................] - 

KeyboardInterrupt: 

## Stop training:

In [22]:
final_model.evaluate(x_test,y_test_ohe)

313/313 [==============================] - 439s 1s/step - loss: 0.2714 - top_1_categorical_accuracy: 0.9167 - top_2_categorical_accuracy: 0.9710 - top_3_categorical_accuracy: 0.9878 - top_4_categorical_accuracy: 0.9949 - top_5_categorical_accuracy: 0.9980 - top_6_categorical_accuracy: 0.9992 - top_7_categorical_accuracy: 0.9995 - top_8_categorical_accuracy: 0.9996 - top_9_categorical_accuracy: 1.0000 - precision_1: 0.9278 - recall_1: 0.9079 - accuracy: 0.0000e+00


[0.2713839113712311,
 0.916700005531311,
 0.9710000157356262,
 0.9878000020980835,
 0.9948999881744385,
 0.9980000257492065,
 0.9991999864578247,
 0.9994999766349792,
 0.9995999932289124,
 1.0,
 0.927848756313324,
 0.9078999757766724,
 0.0]

## top-1 test set categorical accuracy # 0.9167

In [25]:
final_model.save('2022-02-09_manually_trained_exported_final_model')

2022-02-10 07:23:28.949034: W tensorflow/python/util/util.cc:348] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.


INFO:tensorflow:Assets written to: 2022-02-09_manually_trained_exported_final_model/assets


/usr/local/lib/python3.8/dist-packages/keras/utils/generic_utils.py:494: CustomMaskWarning: Custom mask layers require a config and must override get_config. When loading, the custom mask layer must be passed to the custom_objects argument.
  warnings.warn('Custom mask layers require a config and must override '
